src/export_json.lean

```diff
@@ -340,20 +340,50 @@ do docs ← olean_doc_strings,
      (filename, json.array $ l.map write_module_doc_pair) :: rest
    end) [])
 
-meta def get_instances : tactic (rb_lmap string string) :=
+/-- Extract `[foo, bar]` from `has_pow foo bar`.
+
+The result is a map containing the type names as keys.  -/
+meta def find_instance_types (e : expr) : tactic (rb_set string) :=
+fold_explicit_args e mk_rb_set $ λ m e, do
+  e ← whnf e transparency.reducible,
+  t ← infer_type e,
+  -- only look at arguments which are types or propositions
+  expr.sort l ← whnf t transparency.reducible | pure m,
+  some s ← match e.get_app_fn with
+  | expr.const type_name l   := pure (some type_name.to_string)
+  | expr.pi _ _ _ _          := pure (some "pi")
+  | expr.sort level.zero     := pure (some "Prop")
+  | expr.sort (level.succ l) := pure (some "Type")
+  | expr.sort l              := pure (some "Sort")
+  | expr.local_const _ _ _ _ := pure none
+  | expr.macro _ _           := pure none -- TODO: unfold macros?
+  | expr.lam _ _ _ _         := fail ("lam, not a constant: " ++ to_string e)
+  | expr.var i               := fail ("var, not a constant: " ++ to_string e)
+  | expr.mvar _ _ _          := fail ("mvar, not a constant: " ++ to_string e)
+  | expr.app _ _             := fail ("app, not a constant: " ++ to_string e)
+  | expr.elet _ _ _ _        := fail ("elet, not a constant: " ++ to_string e)
+  end | pure m,
+  pure (m.insert s)
+
+meta def get_instances : tactic (rb_lmap string string × rb_lmap string string) :=
 attribute.get_instances `instance >>= list.mfoldl
-  (λ map inst_nm,
+  (λ ⟨map, rev_map⟩ inst_nm,
    do ty ← mk_const inst_nm >>= infer_type,
-      (_, e) ← open_pis_whnf ty transparency.reducible,
+      (binders, e) ← open_pis_whnf ty transparency.reducible,
       e ← whnf e transparency.reducible,
       expr.const class_nm _ ← pure e.get_app_fn |
         fail ("not a constant: " ++ to_string e),
```

Comment:
@gebner, note that we already had a hard failure here anyway.

In my opinion getting the github email is a good thing; upon receipt, we change the hard failure to a TODO comment, but at least then we know what the problem is.

Reply:
> In my opinion getting the github email is a good thing

The problem is that I'm personally getting spammed by these emails every few hours.  Until I either fix it or add an issue (and continue being spammed).  And in the meanwhile the docs are not being updated.

If we can change it so that you get the email and are responsible for fixing the issues, then I'd be happy with adding such easily breaking code.

As far as I'm concerned, the code in this function is purely a heuristic anyhow.  The most likely failure scenario is that someone on Zulip complains about an being included that shouldn't be (see e.g. implicit arguments) or not being included while it should be (see e.g. coercions).  This check won't help with that at all.

Reply:
I've gone ahead and made both failures report to stderr instead.

Reply:
Thanks!